### ISSUE TITLE
API v2

### ISSUE THREAD

**Issue by chrigl** [digitalocean/go-netbox#9]
Is anyone working on the porting go-netbox to APIv2?
I would have time to start on it.

**Comment by mdlayher** [comment]
If you'd like to get started, that would be awesome. Do you have any ideas for a proposed Go API? I am mobile but can share some later toda. I'd like to make sure some consensus is reached before code is written.

**Comment by mdlayher** [comment]
My initial thought is to extend the existing "services" pattern API with an additional service for each API type, such as "aggregate", "IP address", "RIR", etc.

```go
type Client struct {
	IPAM *IPAMService
}

type IPAMService struct {
	c *Client
}

type AggregatesService struct {
	c *Client
}

type Aggregate struct{}

func (s *AggregatesService) List() ([]*Aggregate, error) {
	return []*Aggregate{}, nil
}

func (s *AggregatesService) Get(id int) (*Aggregate, error) {
	return &Aggregate{}, nil
}

func (s *AggregatesService) Create(a *Aggregate) error {
	return nil
}
```

Then it would be called like:

```go
c := netbox.NewClient()
aggrs, _ := c.IPAM.Aggregates.List()
aggr := c.IPAM.Aggregates.Get(1)
c.IPAM.Aggregates.Create(&Aggregate{ID: 1})
```

What do you think?  This way, it's easy to partition up the method set for each type and provide CRUD functionality.

**Comment by mdlayher** [comment]
Oh, and there's also the matter of dealing with pagination now.  I have some ideas in that area but nothing concrete yet.

**Comment by chrigl** [comment]
Structuring looks good to me.

Pagination: I don't have a strong opinion on this. The gophercloud guys always return a Pager.
See:
* https://github.com/gophercloud/gophercloud/blob/master/pagination/pager.go
* https://github.com/gophercloud/gophercloud/blob/master/openstack/compute/v2/servers/requests.go#L62

It feels a little bit odd to use this, but on the other hand, it is easy for a user to stop at a certain point:
```
err := servers.List(client, nil).EachPage(func (page pagination.Page) (bool, error) {
  s, err := servers.ExtractServers(page)
  if err != nil {
    return false, err
  }

  // Handle the []servers.Server slice.

  // Return "false" or an error to prematurely stop fetching new pages.
  return true, nil
})
```

**Comment by mdlayher** [comment]
Apologies, I meant to reply sooner.  I've been playing around with an API similar to existing stdlib iterators, like `bufio.Scanner`.

```go
package main

import (
        "fmt"
        "log"

        "github.com/mdlayher/gonetbox"
)

func main() {
        c, err := gonetbox.NewClient("https://dev-netbox01.example.com/", nil)
        if err != nil {
                log.Fatal(err)
        }

        page := c.DCIM.ListDevices()
        for page.Next() {
                // second return is *http.Response
                devices, _ := page.Devices()

                for _, d := range devices {
                        fmt.Println(d.Name)
                }
        }

        if err := page.Err(); err != nil {
                log.Fatalf("page: %v", err)
        }
}
```

This is one case where it would be pretty handy if Go had generic programming facilities. :)

I'm debating changing `page.Devices()` to something like `page.Extract(devices)` akin to JSON unmarshaling.  Any thoughts?

**Comment by mdlayher** [comment]
On second thought, while this API is pretty nice, it's rather complicated under the hood because of all of the state that needs to be kept.

I think it might make sense to just do what go-github does and return an object indicating where previous/next results can be found: https://godoc.org/github.com/google/go-github/github#Response.

Then the user should be able to paginate pretty easily.  In the case of this package, we could return `Valuer` objects that are preconfigured with the previous/next page, to make it pretty trivial to keep paging through results if needed.

**Comment by mdlayher** [comment]
Here's my new idea.  It's much less complex to implement, at the cost of a slight burden to the caller.

```go
package main

import (
        "fmt"
        "log"

        "github.com/mdlayher/gonetbox"
)

func main() {
        c, err := gonetbox.NewClient("https://dev-netbox01.nyc2.internal.digitalocean.com/", nil)
        if err != nil {
                log.Fatal(err)
        }

        var options gonetbox.Valuer
        for {
                devices, meta, err := c.DCIM.Devices.List(options)
                if err != nil {
                        log.Fatal(err)
                }
                options = meta.Next

                for _, d := range devices {
                        fmt.Printf("[%04d] %s\n", d.ID, d.Name)
                }

                // No next page?
                if meta.Next == nil {
                        break
                }
        }
}
```

**Comment by chrigl** [comment]
I like your first idea, and thought about something like this, while I was using database/sql this week.

I need to play around with it quite a bit, since I have a rough idea how to implement this. Since I already ported ipaddress related structs to v2, I will come back with POC for this.

**Comment by mdlayher** [comment]
Sounds good to me.  I like the API for the first idea but I ran into a couple of roadblocks:

- May need to either implement a pager type for each struct, or add methods to retrieve each type of struct. Or use ugly type assertion. No good.
- The internal state that was used to create the closure was really messy and hard to follow.

That's why I think I'm leaning toward option 2. Pagination just ends up being providing limit and offset query parameters, and as my example shows, it isn't too bad on the user. We could even wrap it up in a utility function.

Ultimately I'd like to go with something that is simple and easy to maintain, and doesn't subvert Go's type system. I'm happy to look over your PoC when it's ready.

**Comment by chrigl** [comment]
Here comes my approach: I created a branch develop-v2 in my repo. Like I mentioned, I already ported IPAddress (which might not work with tenants! not tested).

First of all, I added a new file ([page.go](https://github.com/chrigl/go-netbox/blob/develop-v2/page.go) and changed `Client.IPAM` to contain a field `IPAddresses *IPAddressesService`. To this I added the Functions `Get`, `List`, and `Extract`. While get works like `GetIPAddress` before, `List` was changed to return a `Page`, which must be traversed by the user. The user is also responsible for calling the "correct" Extract function against `page`. But this is IMHO ok, and I don't need any go-type-voodoo.

Of course, documentation and tests are missing, but I want to hear your thoughts on the implementation, and the resulted API.

I have a working netbox v2 (without configured tenancy) against which the following code works:

```
package main

import (
        "fmt"
        "github.com/digitalocean/go-netbox"
        "net/http"
        "os"
)

type TokenAuthTransport struct {
        Token string
        rt    http.RoundTripper
}

func (b TokenAuthTransport) RoundTrip(req *http.Request) (*http.Response, error) {
        req.Header.Set("Authorization", fmt.Sprintf("Token %s", b.Token))
        if b.rt == nil {
                b.rt = http.DefaultTransport
        }
        return b.rt.RoundTrip(req)
}

// Create a new netbox client with a Token Authentication RoundTripper
func NewClient(token string, url string) (*netbox.Client, error) {
        httpClient := &http.Client{Transport: TokenAuthTransport{Token: token}}
        client, err := netbox.NewClient(url, httpClient)
        if err != nil {
                return nil, err
        }
        return client, nil
}

func main() {
        client, err := NewClient(os.Getenv("NETBOX_TOKEN"), os.Getenv("NETBOX_URL"))
        if err != nil {
                panic(err)
        }

        x, err := client.IPAM.IPAddresses.Get(155)
        fmt.Printf("Got 2: %v\n", x)
        fmt.Printf("Status: %v\n", x.Status)

        i := 0
        page := client.IPAM.IPAddresses.List(nil)
        for page.Next() {
                res, err := client.IPAM.IPAddresses.Extract(page)
                if err != nil {
                        fmt.Println(err)
                        break
                }
                for _, ip := range res {
                        i = i + 1
                        fmt.Printf("%d. %s\n", i, ip.Address)
                }
        }
        if err := page.Err(); err != nil {
                panic(err)
        }
}
```

**Comment by mdlayher** [comment]
Looks great to me!  If you'd like to start submitting pull requests, I'd be happy to have them.

**Comment by chrigl** [comment]
One more thing, I am not sure about.

How do you like to have the changes. I very likely need to break some go-api, because of changes in the types. Should we keep v1 in the code and v2 parallel to it? If so, should we create sub-package e.g. `apiv2`, with completely own types etc?
Or should we do a hard cut. Port everything to v2 and throw away v1, and hope everyone has vendored the used version of go-netbox?

**Comment by mdlayher** [comment]
I think I like the last option.  While we're moving things around, I'd also like to make the package name actually line up with the directory name (just `netbox` and not `go-netbox`).

Perhaps we can do:
- `github.com/digitalocean/go-netbox/netboxv1`
- `github.com/digitalocean/go-netbox/netboxv2`

All legacy code gets moved to `netboxv1` and new code goes in `netboxv2`.

Or we just throw away all v1 code and do `github.com/digitalocean/go-netbox/netbox`.

@jeremystretch tells me that the 1.0 API is gone in NetBox 2.0, so I'm fine with just tossing that code.

What do you think?

**Comment by chrigl** [comment]
I like `github.com/digitalocean/go-netbox/netbox`. As long as we are working on v2, we could keep v1 in its current form. So master will be still usable, until we decide to finally throw v1 away.

**Comment by mdlayher** [comment]
Sounds good to me!  I'd like to review the code in smaller sections if possible though, just so it doesn't end up being a huge PR.

Any suggestions on how to do that?  I suppose I could create a small 2.0 branch here and have you send PRs to it.

**Comment by chrigl** [comment]
Sounds good to me.

I will start with the paging and client, since these are the building block for almost everything else. Then I try to start with self contained blocks like ipam_vrf, ipam_roles, tenancy, dcim_site, etc, slowly moving up to the large composed types like ipam_ipaddresses. I think this results in small chunks to merge.

**Comment by mdlayher** [comment]
Sounds good to me as well.  I just created branch `v2` here upstream.  If you'd like to rebase and start submitting PRs against that, that would be awesome.  I know that some of the v1 code will be copied as-is, but I figure it's best to just start fresh and move things over as needed.

**Comment by mdlayher** [comment]
Oh, one more thing: don't worry about implementing every single API endpoint by yourself.  It's a drag and it'll take ages.

Implement what you need most, and we can try to lean on others (and myself) to add the missing pieces.

**Comment by adamdecaf** [comment]
I'm looking at writing a terraform provider for netbox (we might use it at work) and would be willing to write endpoints if needed. Right now I'm still learning about netbox and figuring out what terraform would help with. 

Initially we'll probably want to use the IPAM portions of netbox.

**Comment by chrigl** [comment]
@adamdecaf Currently I am working on a PR for the first endpoints. If this is done, we could start to create more.

The endpoint I currently need most is ip-addresses, and with it devices and interfaces. So this might be the next ones I will work on.

**Comment by d0m0reg00dthing** [comment]
I have sometime working on this project. May I join this project, I'm not a developer but I think I could learn how to code in go.

**Comment by d0m0reg00dthing** [comment]
And if yes, Should I fork from digitalocean or chrigl repository to start working

**Comment by mdlayher** [comment]
You'll want to fork this repository, but I recommend letting us land #13 first since there's a lot going on.

**Comment by d0m0reg00dthing** [comment]
I think chrigl should create PR for his work done, many endpoint type depending on paging, vrf and tenancy that he is working on

**Comment by chrigl** [comment]
There is #13 which includes a tiny code generator to make our lives easier. I will come back to this tomorrow and resolve the remaining comments. With this merged, we can start to create all the types and endpoints.

**Comment by d0m0reg00dthing** [comment]
I was clone chrigl repository and do some thing related to Region. Get/List is ok now but when Create/Update Region, I unable to find where to insert Token to it. Any idea?

**Comment by chrigl** [comment]
You need to pass a prepared `http.Client` to go-netbox. We should document this in the README.

Something like
```
type TokenAuthTransport struct {
        Token string
        rt    http.RoundTripper
}

func (b *TokenAuthTransport) RoundTrip(req *http.Request) (*http.Response, error) {
        req.Header.Set("Authorization", fmt.Sprintf("Token %s", b.Token))
        if b.rt == nil {
                b.rt = http.DefaultTransport
        }
        return b.rt.RoundTrip(req)
}

func newClient(token string, url string) (*netbox.Client, error) {
        httpClient := &http.Client{Transport: TokenAuthTransport{Token: token}}
        client, err := netbox.NewClient(url, httpClient)
        if err != nil {
                return nil, err
        }
        return client, nil
}
```
